pkg/sql/logictest/logic.go

```diff
@@ -3431,6 +3546,24 @@ func (t *logicTest) processSubtest(
 			}
 			t.traceStop()
 
+		case "upgrade":
+			if len(fields) != 2 {
+				return errors.Errorf("traceon requires a filename argument, found: %v", fields)
+			}
+			if t.testserverCluster == nil {
+				return errors.Errorf("not a cockroach-go/testserver cluster")
+			}
+			nodeNum, err := strconv.Atoi(fields[1])
+			if err != nil {
+				t.Fatal(err)
+			}
+			if err := t.testserverCluster.UpgradeNode(nodeNum); err != nil {
```

Comment:
The startup logic is using fixed ports [1] which means we can't run these in parallel; even running them sequentially would yield flakes due to (non-deterministic) port reuse.

[1] https://github.com/cockroachdb/cockroach-go/blob/master/testserver/testserver.go#L457